contracts/starknet/space/space.cairo

```diff
@@ -52,56 +62,124 @@ end
+    if _voting_strategies_len == 0:
+        # List is empty
+        return ()
+    else:
+        # Add voting strategy
+        voting_strategies.write(_voting_strategies[0], 1)
+
+        if _voting_strategies_len == 1:
+            # Nothing left to add, end recursion
+            return ()
+        else:
+            # Recurse
+            register_voting_strategies(_voting_strategies_len - 1, &_voting_strategies[1])
+            return ()
+        end
+    end
+end
+
+func register_authenticators{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(
+        _authenticators_len : felt, _authenticators : felt*):
+    if _authenticators_len == 0:
+        # List is empty
+        return ()
+    else:
+        # Add voting strategy
+        authenticators.write(_authenticators[0], 1)
+
+        if _authenticators_len == 1:
+            # Nothing left to add, end recursion
+            return ()
+        else:
+            # Recurse
+            register_authenticators(_authenticators_len - 1, &_authenticators[1])
+            return ()
+        end
+    end
+end
+
 @constructor
 func constructor{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr : felt}(
         _voting_delay : felt, _voting_period : felt, _proposal_threshold : Uint256,
-        _voting_strategy : felt, _authenticator : felt):
+        _executor : felt, _voting_strategies_len : felt, _voting_strategies : felt*,
+        _authenticators_len : felt, _authenticators : felt*):
     # Sanity checks
     assert_nn(_voting_delay)
     assert_nn(_voting_period)
-    assert_not_zero(_voting_strategy)
-    assert_not_zero(_authenticator)
+    assert_not_zero(_voting_strategies_len)
+    assert_not_zero(_authenticators_len)
     # TODO: maybe use uint256_signed_nn to check proposal_threshold?
+    # TODO: maybe check that _executor is not 0?
 
     # Initialize the storage variables
     voting_delay.write(_voting_delay)
     voting_period.write(_voting_period)
     proposal_threshold.write(_proposal_threshold)
-    voting_strategy.write(_voting_strategy)
-    authenticator.write(_authenticator)
+    executor.write(_executor)
+
+    register_voting_strategies(_voting_strategies_len, _voting_strategies)
+    register_authenticators(_authenticators_len, _authenticators)
+
     next_proposal_nonce.write(1)
 
     return ()
 end
 
 @external
 func vote{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr : felt}(
-        voter_address : EthAddress, proposal_id : felt, choice : felt, params_len : felt,
-        params : felt*) -> ():
+        voting_strategy_contract : felt, voter_address : EthAddress, proposal_id : felt,
+        choice : felt, voting_params_len : felt, voting_params : felt*) -> ():
     alloc_locals
 
     # Verify that the caller is the authenticator contract.
-    authenticator_only()
+    assert_valid_authenticator()
+    assert_valid_voting_strategy(voting_strategy_contract)
```

Comment:
The way I saw it, There is a set of whitelisted voting strategies set in the space and when you vote, `get_voting_power` will be called on each strategy in the list.

Reply:
thanks! 8a221a86c31608186f1bdd54c0c7388e88878beb

Reply:
thanks! 8a221a86c31608186f1bdd54c0c7388e88878beb

Reply:
> The way I saw it, There is a set of whitelisted voting strategies set in the space and when you vote, `get_voting_power` will be called on each strategy in the list. The overall voting power for the voter will be the sum of voting powers from each strategy. voters do not get to choose what strategy they use

We've talked about it yesterday on Discord. For future reference, we've decided that:
1) If an overflow happens (or any other logic error happens while voting), we don't automatically cancel the proposal. It is up to the `controller` to decide what to do in this scenario.
2) We will however need to implement a `cancel_proposal` function, only callable by the `controller`, which will basically be like `finalize_proposal`. This `cancel_proposal` should call the execution contract, and we should rename `hasPassed` to something like `proposalStatus` or `proposalOutcomeStatus`, and have different status `ACCEPTED, REJECTED, CANCELLED`. This will allow execution strategies to handle the logic differently :)

Reply:
> The way I saw it, There is a set of whitelisted voting strategies set in the space and when you vote, `get_voting_power` will be called on each strategy in the list. The overall voting power for the voter will be the sum of voting powers from each strategy. voters do not get to choose what strategy they use

We've talked about it yesterday on Discord. For future reference, we've decided that:
1) If an overflow happens (or any other logic error happens while voting), we don't automatically cancel the proposal. It is up to the `controller` to decide what to do in this scenario.
2) We will however need to implement a `cancel_proposal` function, only callable by the `controller`, which will basically be like `finalize_proposal`. This `cancel_proposal` should call the execution contract, and we should rename `hasPassed` to something like `proposalStatus` or `proposalOutcomeStatus`, and have different status `ACCEPTED, REJECTED, CANCELLED`. This will allow execution strategies to handle the logic differently :)

Reply:
> The way I saw it, There is a set of whitelisted voting strategies set in the space and when you vote, `get_voting_power` will be called on each strategy in the list. The overall voting power for the voter will be the sum of voting powers from each strategy. voters do not get to choose what strategy they use

We've talked about it yesterday on Discord. For future reference, we've decided that:
1) If an overflow happens (or any other logic error happens while voting), we don't automatically cancel the proposal. It is up to the `controller` to decide what to do in this scenario.
2) We will however need to implement a `cancel_proposal` function, only callable by the `controller`, which will basically be like `finalize_proposal`. This `cancel_proposal` should call the execution contract, and we should rename `hasPassed` to something like `proposalStatus` or `proposalOutcomeStatus`, and have different status `ACCEPTED, REJECTED, CANCELLED`. This will allow execution strategies to handle the logic differently :)

Reply:
implemented in https://github.com/snapshot-labs/sx-core/pull/67/commits/e41d8bf32a0fef85620b8bec941c46d50d8b10fb